# CSV preprocessing
- 결측치의 확인
- 좋음, 나쁨 등의 데이터 라벨링

In [21]:
import pandas as pd

In [22]:
restaurant_df = pd.read_csv("db_csv/restaurant_df.csv")
menu_df = pd.read_csv("db_csv/menu_df.csv")
review_df = pd.read_csv("db_csv/review_df.csv")

## 결측치 확인

In [23]:
restaurant_df.head()

,restaurant_code,restaurant_name,img_link,region,category,address,open_time,close_time,tel_no,tags,character
0,0,유태우스시,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"스시,회전초밥",서울특별시 동작구 보라매로 113 1층 유태우스시,11:00,22:00,0507-1358-1246,"데이트,혼밥","가성비좋은,서민적인,점심식사,저녁식사,바테이블,다찌석"
1,1,듀윗,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"디저트,케이크",서울특별시 동작구 대방동1길 10 1층,08:00,21:00,0507-2093-6672,"데이트,기념일","예쁜,깔끔한,점심식사,테이크아웃,예약가능"
2,2,한제소곱창전골 신대방삼거리역점,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"곱창전골,소곱창전골","서울특별시 동작구 여의대방로22길 121 상가동 1층 1444호, 1445호",16:00,00:00,0507-1359-9161,"술모임,해장용","지역주민이찾는,매콤한,저녁식사,예약가능,배달"
3,3,프로스퍼,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"카페,커피",서울특별시 동작구 여의대방로24길 101 1층,10:00,00:00,0507-1361-3276,"차모임,혼자카페","캐주얼한,카공,점심식사,테이크아웃,와이파이"
4,4,일진아구찜,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,아구찜,서울특별시 동작구 여의대방로24길 93 1층,11:00,22:00,02-817-5554,가족외식,"서민적인,매콤한,점심식사,저녁식사,유아의자"


In [24]:
print(len(restaurant_df))
restaurant_df.isna().sum()

100


restaurant_code     0
restaurant_name     0
img_link            0
region              0
category            0
address             0
open_time           9
close_time          9
tel_no              5
tags               12
character           2
dtype: int64

In [25]:
na_series = restaurant_df["open_time"].isna()
na_series = na_series | restaurant_df["close_time"].isna()
na_series = na_series | restaurant_df["tel_no"].isna()
na_series = na_series | restaurant_df["tags"].isna()
na_series = na_series | restaurant_df["character"].isna()
print(na_series.sum())
restaurant_df[na_series]

19


,restaurant_code,restaurant_name,img_link,region,category,address,open_time,close_time,tel_no,tags,character
14,14,원조부안집 신대방삼거리역점,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"고기집,목살",서울특별시 동작구 상도로12길 20 1층,14:00,01:00,0507-1495-8314,NaN,"서민적인,국내산,저녁식사,연탄불,무료주차"
20,20,먹거리장터,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리,포차,서울특별시 동작구 여의대방로24차길 48 샬롬마트,NaN,NaN,02-812-1101,"술모임,간단하게한잔","서민적인,2차,바깥자리"
36,36,파도회,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리,"횟집,포차",서울특별시 동작구 신대방동 여의대방로24차길 48,NaN,NaN,02-821-5576,술모임,"서민적인,야장,야외자리,예약불가"
38,38,홍가네반찬,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"게장,양념게장",서울특별시 동작구 국사봉1길 45,08:00,20:00,02-822-7788,NaN,"서민적인,다양한반찬,배달"
44,44,우리숯불국밥 본정,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,국밥,"서울특별시 동작구 국사봉1길 12 ,1층(상도동, 진선빌딩)",NaN,NaN,NaN,NaN,NaN
46,46,소문난포장마차,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리,포차,서울특별시 동작구 여의대방로24차길 48,NaN,NaN,02-813-2236,"술모임,낮술","서민적인,야장,야외자리"
48,48,라밥 신대방삼거리점,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"밥집,김밥",서울특별시 동작구 보라매로 106 1층,00:00,24:00,02-821-5400,NaN,"깔끔한,다양한메뉴,점심식사,배달"
50,50,고유부,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리,"묵은지,김밥",서울특별시 동작구 국사봉길 33 1층,08:00,21:00,0507-1410-7819,NaN,"푸짐한,동네사람,아침식사,점심식사,저녁식사,배달"
54,54,은우네간식,https://d12zq4w4guyljn.cloudfront.net/750_750_...,신대방삼거리역,붕어빵,서울특별시 동작구 국사봉1길 63,NaN,NaN,NaN,NaN,NaN
65,65,한울전집,https://d12zq4w4guyljn.cloudfront.net/750_750_...,신대방삼거리역,전집,서울특별시 동작구 국사봉1길 37 2층,16:00,01:00,02-824-9020,NaN,"비오는날,안주"


개점, 폐점 시간, 전화번호, 태그, 특징(character) 기입하지 않은 매장이 존재

100개 매장 중 19개로 많은 편

In [26]:
menu_df.head()

,restaurant_code,menu_name,price,description
0,0,회전초밥,NaN,NaN
1,0,장어초밥,NaN,NaN
2,1,프레지에 1호 홀케이크,"50,000원",NaN
3,1,프레지에,"10,500원",NaN
4,1,페르디셔 압펠,"6,000원",NaN


In [27]:
print(len(menu_df))
menu_df.isna().sum()

2008


restaurant_code       0
menu_name             0
price                19
description        1028
dtype: int64

In [28]:
no_price_df = menu_df[menu_df["price"].isna()]
no_price_df

,restaurant_code,menu_name,price,description
0,0,회전초밥,NaN,NaN
1,0,장어초밥,NaN,NaN
416,14,10분밥,NaN,"좋은 쌀로 갓 지은 밥으로 버터, 계란 , 참기름, 등을 비벼 먹는 시그니처 메뉴"
474,20,두부김치,NaN,NaN
475,20,부추전,NaN,NaN
476,20,잔치국수,NaN,NaN
890,38,양념게장,NaN,NaN
891,38,나물,NaN,NaN
893,38,잡채,NaN,NaN
894,38,전,NaN,NaN


In [29]:
res = no_price_df["restaurant_code"].unique()
restaurant_df[restaurant_df["restaurant_code"].isin(res)]

,restaurant_code,restaurant_name,img_link,region,category,address,open_time,close_time,tel_no,tags,character
0,0,유태우스시,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"스시,회전초밥",서울특별시 동작구 보라매로 113 1층 유태우스시,11:00,22:00,0507-1358-1246,"데이트,혼밥","가성비좋은,서민적인,점심식사,저녁식사,바테이블,다찌석"
14,14,원조부안집 신대방삼거리역점,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"고기집,목살",서울특별시 동작구 상도로12길 20 1층,14:00,01:00,0507-1495-8314,NaN,"서민적인,국내산,저녁식사,연탄불,무료주차"
20,20,먹거리장터,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리,포차,서울특별시 동작구 여의대방로24차길 48 샬롬마트,NaN,NaN,02-812-1101,"술모임,간단하게한잔","서민적인,2차,바깥자리"
38,38,홍가네반찬,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"게장,양념게장",서울특별시 동작구 국사봉1길 45,08:00,20:00,02-822-7788,NaN,"서민적인,다양한반찬,배달"
39,39,계림원,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"통닭,누룽지통닭",서울특별시 동작구 상도로12길 13,12:00,23:30,02-826-8996,"단체모임,모임장소","야장,넓은,점심식사,저녁식사,테라스,야외테라스"
51,51,뿌루밀리 신대방삼거리역점,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"커피,테이크아웃카페",서울특별시 동작구 국사봉길 2 1층 뿌루밀리,08:00,22:00,02-3280-3330,간식,"캐주얼한,저렴한,점심식사,테이크아웃"
66,66,시래불고기 화풍정,https://d12zq4w4guyljn.cloudfront.net/300_300_...,성대시장,"시래기불고기,한식",서울특별시 동작구 상도로 102 성대전통시장 2층,11:00,21:30,0507-1489-3507,"모범음식점,가족외식,회식","서민적인,깔끔한,점심식사,저녁식사,무료주차"
93,93,성대낙지,https://d12zq4w4guyljn.cloudfront.net/300_300_...,신대방삼거리역,"낙지,낙지비빔밥",서울특별시 동작구 국사봉1길 4,11:30,22:00,02-816-4347,혼밥,"서민적인,매콤한,점심식사,연중무휴"


확인 결과 해당 음식점들은 전부, 혹은 일부 메뉴의 가격 정보를 기입하지 않은 상태였음

2008개의 메뉴 중 19개의 메뉴로 검색시 이를 고려해서 검색 결과를 보여줄 필요가 있어보임

In [30]:
review_df.head()

,restaurant_code,reviewer_name,reviewer_avg_score,reviewer_rv_cnt,reviewer_follower_cnt,reviewer_score,taste_level,price_level,service_level,review_content,review_menu,keywords
0,0,냐밍,4.3,355,69,4점,좋음,만족,보통,요즘 회전초밥집 정말 비싼데 여긴 가성비 갑이었어요! 위치는 조금 복잡한 곳에 있긴...,회전초밥,"저녁식사,식사모임,캐주얼한,이국적/이색적,가성비좋은"
1,0,발통이,4.5,113,1,4점,보통,만족,좋음,한접시에 1500원이면 저렴해요 근데 맛도 생각보다 괜찮았어요! 엄청난 퀄리티는 아...,스시,"점심식사,지역주민이찾는"
2,0,우리의나날,4.5,420,2,5점,좋음,만족,좋음,생일날 갔어요. 저녁시간 웨이팅있었지만 금방 자리에 앉았어요. 회전초밥 모든접시가 ...,회전초밥 1500원,"저녁식사,데이트,서민적인,푸짐한,지역주민이찾는"
3,0,냠냠냠히,4.8,71,0,5점,좋음,만족,좋음,"모든접시 1,700원 이라는 가성비에 맛도 좋고 이런 굿굿!!","연어,장어","저녁식사,가성비좋은"
4,0,밍민,4.8,232,2,5점,좋음,만족,좋음,1500원이라가성비좋아요 다양하게 먹을 수 있고 여러가지종류있어서 좋아요,회전초밥,"점심식사,캐주얼한"


In [31]:
print(len(review_df))
review_df.isna().sum()

422


restaurant_code          0
reviewer_name            0
reviewer_avg_score       0
reviewer_rv_cnt          0
reviewer_follower_cnt    0
reviewer_score           0
taste_level              7
price_level              7
service_level            7
review_content           7
review_menu              8
keywords                 7
dtype: int64

In [32]:
na_series = review_df["taste_level"].isna()
na_series = na_series | review_df["price_level"].isna()
na_series = na_series | review_df["service_level"].isna()
na_series = na_series | review_df["review_content"].isna()
na_series = na_series | review_df["review_menu"].isna()
na_series = na_series | review_df["keywords"].isna()
print(na_series.sum())
review_df[na_series]

12


,restaurant_code,reviewer_name,reviewer_avg_score,reviewer_rv_cnt,reviewer_follower_cnt,reviewer_score,taste_level,price_level,service_level,review_content,review_menu,keywords
31,1,다코미식가\nAriel,3.8,91,27,4점,좋음,보통,좋음,"가격이 비싼만큼 고급진 분위기, 커피랑 디저트 맛있어용❤️",NaN,"디저트,데이트,여자끼리,조용한,아담한,고급스러운,모던한"
32,1,Jisu Pyo,4.3,2,0,3점,NaN,NaN,NaN,디저트는 나쁘지 않은데 홍차가 아주 별로였어요. 포트에 잎을 너무 적게 넣어주어서 ...,NaN,NaN
53,4,손미정,3.9,15,0,5점,NaN,NaN,NaN,어릴 적 부터 갔는데 반찬 맛도 아구찜 맛도 변치 않고 맛있음\n그 전 같은거랑 오...,NaN,NaN
93,9,만두,5.0,2,0,5점,NaN,NaN,NaN,NaN,NaN,NaN
144,16,손미정,3.9,15,0,5점,NaN,NaN,NaN,가끔 진짜 미친듯이 여기 칼국수 먹고싶음 ㅠㅠ,NaN,NaN
162,19,Sang Jin Hong,4.3,3,0,4점,NaN,NaN,NaN,술 마시고 집에 들어가기 아쉬울때.,NaN,NaN
163,19,장재호,4.0,3,0,4점,NaN,NaN,NaN,NaN,NaN,NaN
183,24,덕구K,5.0,1,0,5점,보통,만족,보통,NaN,사이야 이자카이,"술모임,고급스러운,조용한,이국적/이색적,무료주차"
289,52,MinChae Kang,5.0,1,0,5점,NaN,NaN,NaN,NaN,NaN,NaN
356,74,블룸39,4.9,7,1,5점,좋음,만족,좋음,NaN,"안심 돈카츠(히레),특 모둠 카츠,치즈롤 돈카츠","아이동반,점심식사,저녁식사,데이트,가족외식,숨은맛집,고급스러운,가성비좋은,조용한,깔..."


맛, 가격, 서비스 수준을 응답하지 않은 사람

리뷰 내용이 공백인 사람

먹은 메뉴가 무엇인지 기입하지 않은 사람

키워드를 적지 않은 사람

422명 중 도합 12명 존재

## 데이터 라벨링

In [33]:
# menu_df의 price 숫자로 바꾸기
import re
def parse_price(instr):
    if pd.isna(instr):
        return ""

    digits = re.findall(r"\d+", instr)
    return "".join(digits)

menu_df["price"] = menu_df["price"].apply(parse_price)
menu_df.head()

,restaurant_code,menu_name,price,description
0,0,회전초밥,,NaN
1,0,장어초밥,,NaN
2,1,프레지에 1호 홀케이크,50000,NaN
3,1,프레지에,10500,NaN
4,1,페르디셔 압펠,6000,NaN


In [34]:
# 데이터 라벨링을 위한 고유값 확인
print(review_df["taste_level"].unique()) # 부족, 보통, 좋음
print(review_df["price_level"].unique()) # 불만, 보통, 만족
print(review_df["service_level"].unique()) # 나쁨, 보통, 좋음

['좋음' '보통' nan '부족']
['만족' '보통' '불만' nan]
['보통' '좋음' nan '나쁨']


In [ ]:
taste_labels = ['부족', '보통', '좋음']
price_labels = ['불만', '보통', '만족']
service_labels = ['나쁨', '보통', '좋음']

target_labels = []
def lebelize(instr):
    if pd.isna(instr):
        return ""
    
    if instr not in target_labels:
        return instr

    return target_labels.index(instr)

def parse_score(instr):
    return float(instr[:-1])

target_labels = taste_labels
review_df["taste_level"] = review_df["taste_level"].apply(lebelize)
target_labels = price_labels
review_df["price_level"] = review_df["price_level"].apply(lebelize)
target_labels = service_labels
review_df["service_level"] = review_df["service_level"].apply(lebelize)
review_df["reviewer_score"] = review_df["reviewer_score"].apply(parse_score)
review_df.head()


,restaurant_code,reviewer_name,reviewer_avg_score,reviewer_rv_cnt,reviewer_follower_cnt,reviewer_score,taste_level,price_level,service_level,review_content,review_menu,keywords
0,0,냐밍,4.3,355,69,4점,2,2,1,요즘 회전초밥집 정말 비싼데 여긴 가성비 갑이었어요! 위치는 조금 복잡한 곳에 있긴...,회전초밥,"저녁식사,식사모임,캐주얼한,이국적/이색적,가성비좋은"
1,0,발통이,4.5,113,1,4점,1,2,2,한접시에 1500원이면 저렴해요 근데 맛도 생각보다 괜찮았어요! 엄청난 퀄리티는 아...,스시,"점심식사,지역주민이찾는"
2,0,우리의나날,4.5,420,2,5점,2,2,2,생일날 갔어요. 저녁시간 웨이팅있었지만 금방 자리에 앉았어요. 회전초밥 모든접시가 ...,회전초밥 1500원,"저녁식사,데이트,서민적인,푸짐한,지역주민이찾는"
3,0,냠냠냠히,4.8,71,0,5점,2,2,2,"모든접시 1,700원 이라는 가성비에 맛도 좋고 이런 굿굿!!","연어,장어","저녁식사,가성비좋은"
4,0,밍민,4.8,232,2,5점,2,2,2,1500원이라가성비좋아요 다양하게 먹을 수 있고 여러가지종류있어서 좋아요,회전초밥,"점심식사,캐주얼한"


In [36]:
(review_df == "").sum()

restaurant_code          0
reviewer_name            0
reviewer_avg_score       0
reviewer_rv_cnt          0
reviewer_follower_cnt    0
reviewer_score           0
taste_level              7
price_level              7
service_level            7
review_content           0
review_menu              0
keywords                 0
dtype: int64

## 기타 분석

In [37]:
print(len(review_df))
print(len(review_df["reviewer_name"].unique()))

422
171


422개 리뷰가 171명에 의해서 작성됨

리뷰 유저도 다른 table로 작성 가능해 보임

## 결측치 최종 처리

In [38]:
# 남은 결측치는 일단 공백 처리
restaurant_df.fillna("")
menu_df.fillna("")
review_df.fillna("")

,restaurant_code,reviewer_name,reviewer_avg_score,reviewer_rv_cnt,reviewer_follower_cnt,reviewer_score,taste_level,price_level,service_level,review_content,review_menu,keywords
0,0,냐밍,4.3,355,69,4점,2,2,1,요즘 회전초밥집 정말 비싼데 여긴 가성비 갑이었어요! 위치는 조금 복잡한 곳에 있긴...,회전초밥,"저녁식사,식사모임,캐주얼한,이국적/이색적,가성비좋은"
1,0,발통이,4.5,113,1,4점,1,2,2,한접시에 1500원이면 저렴해요 근데 맛도 생각보다 괜찮았어요! 엄청난 퀄리티는 아...,스시,"점심식사,지역주민이찾는"
2,0,우리의나날,4.5,420,2,5점,2,2,2,생일날 갔어요. 저녁시간 웨이팅있었지만 금방 자리에 앉았어요. 회전초밥 모든접시가 ...,회전초밥 1500원,"저녁식사,데이트,서민적인,푸짐한,지역주민이찾는"
3,0,냠냠냠히,4.8,71,0,5점,2,2,2,"모든접시 1,700원 이라는 가성비에 맛도 좋고 이런 굿굿!!","연어,장어","저녁식사,가성비좋은"
4,0,밍민,4.8,232,2,5점,2,2,2,1500원이라가성비좋아요 다양하게 먹을 수 있고 여러가지종류있어서 좋아요,회전초밥,"점심식사,캐주얼한"
...,...,...,...,...,...,...,...,...,...,...,...,...
417,96,안쓰러운샐러드소스,5.0,117,1,5점,2,2,2,저희 동네 명물이에요.. 너무 맛있어요 특히 호떡!!! 꼭 드셔보세요!!,"닭꼬치,흑미호떡,어묵","아이동반,데이트,간식,숨은맛집,서민적인,지역주민이찾는"
418,96,먼지님,4.5,532,47,5점,2,2,2,우리동네 오뎅 맛집입니다\n오뎅국물이 넘 맛나요\n겨울에는 자주갑니다.,오뎅,"저녁식사,서민적인,시끌벅적한,주차불가"
419,97,야나두,3.6,376,4,5점,2,2,2,노부부가 운영하시는 새로생긴 붕어빵집인데 여기 붕어빵이 공장처럼 기계식으로 만들어져...,팥붕어빵 2개,"간식,시끌벅적한"
420,98,다닝12,4.4,424,12,4점,2,1,2,"캔에 담아주셔서 대중교통 이용하기 편함. \n라떼 4,500원으로 가격은 저렴한 편...",아이스라떼,"혼카페,테이크아웃,숨은맛집,깔끔한"


In [39]:
restaurant_df.to_csv("db_csv_preprocessed/restaurant_df.csv", index=False)
menu_df.to_csv("db_csv_preprocessed/menu_df.csv", index=False)
review_df.to_csv("db_csv_preprocessed/review_df.csv", index=False)